### PR TITLE
config-loader: reject invalid configuration schema

### LINKED ISSUE #18812: 🐛 Bug Report: Secret not redacted in DevTools Config tab
### 📜 Description

When using the [catalog-backend-module-msgraph](https://github.com/backstage/backstage/tree/dcd7e1622c3b8008f7561ba48f94010672159a48/plugins/catalog-backend-module-msgraph), the following secret, which is correctly annotated in the plugin, is not redacted: config.catalog.providers.microsoftGraphOrg.default.clientSecret

Similar Issue regarding other plugins: https://github.com/backstage/backstage/issues/18176

### 👍 Expected behavior

The secret should be redacted

### 👎 Actual Behavior with Screenshots

image

### 👟 Reproduction steps

Install the plugin and set the mentioned configuration

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

OS: Linux 4.18.0-372.36.1.el8_6.x86_64 - linux/x64
node: v18.17.0
backstage: 1.16.0
Dependencies:
    @backstage/app-defaults: 1.4.1
    @backstage/backend-app-api: 0.4.5
    @backstage/backend-common: 0.19.1
    @backstage/backend-dev-utils: 0.1.1
    @backstage/backend-plugin-api: 0.5.4
    @backstage/backend-tasks: 0.5.4
    @backstage/catalog-client: 1.4.3
    @backstage/catalog-model: 1.4.1
    @backstage/cli-common: 0.1.12
    @backstage/cli-node: 0.1.2
    @backstage/cli: 0.22.9
    @backstage/config-loader: 1.3.2
    @backstage/config: 1.0.8
    @backstage/core-app-api: 1.9.0
    @backstage/core-components: 0.13.3
    @backstage/core-plugin-api: 1.5.3
    @backstage/dev-utils: 1.0.17
    @backstage/errors: 1.2.1
    @backstage/eslint-plugin: 0.1.3
    @backstage/integration-aws-node: 0.1.5
    @backstage/integration-react: 1.1.15
    @backstage/integration: 1.5.1
    @backstage/plugin-api-docs: 0.9.6
    @backstage/plugin-app-backend: 0.3.47
    @backstage/plugin-auth-backend: 0.18.5
    @backstage/plugin-auth-node: 0.2.16
    @backstage/plugin-catalog-backend-module-bitbucket-server: 0.1.12
    @backstage/plugin-catalog-backend-module-msgraph: 0.5.6
    @backstage/plugin-catalog-backend: 1.11.0
    @backstage/plugin-catalog-common: 1.0.15
    @backstage/plugin-catalog-graph: 0.2.32
    @backstage/plugin-catalog-import: 0.9.10
    @backstage/plugin-catalog-node: 1.4.0
    @backstage/plugin-catalog-react: 1.8.0
    @backstage/plugin-catalog: 1.12.0
    @backstage/plugin-devtools-backend: 0.1.2
    @backstage/plugin-devtools-common: 0.1.2
    @backstage/plugin-devtools: 0.1.2
    @backstage/plugin-events-node: 0.2.8
    @backstage/plugin-github-actions: 0.6.1
    @backstage/plugin-home-react: 0.1.1
    @backstage/plugin-home: 0.5.4
    @backstage/plugin-kubernetes-backend: 0.11.2
    @backstage/plugin-kubernetes-common: 0.6.5
    @backstage/plugin-kubernetes: 0.9.3
    @backstage/plugin-org: 0.6.10
    @backstage/plugin-permission-backend: 0.5.22
    @backstage/plugin-permission-common: 0.7.7
    @backstage/plugin-permission-node: 0.7.10
    @backstage/plugin-permission-react: 0.4.14
    @backstage/plugin-proxy-backend: 0.2.41
    @backstage/plugin-scaffolder-backend: 1.15.1
    @backstage/plugin-scaffolder-common: 1.3.2
    @backstage/plugin-scaffolder-node: 0.1.5
    @backstage/plugin-scaffolder-react: 1.5.1
    @backstage/plugin-scaffolder: 1.14.1
    @backstage/plugin-search-backend-module-catalog: 0.1.3
    @backstage/plugin-search-backend-module-pg: 0.5.8
    @backstage/plugin-search-backend-node: 1.2.3
    @backstage/plugin-search-backend: 1.3.3
    @backstage/plugin-search-common: 1.2.5
    @backstage/plugin-search-react: 1.6.3
    @backstage/plugin-search: 1.3.3
    @backstage/plugin-todo-backend: 0.1.44
    @backstage/plugin-todo: 0.2.22
    @backstage/plugin-user-settings: 0.7.5
    @backstage/release-manifests: 0.0.9
    @backstage/test-utils: 1.4.1
    @backstage/theme: 0.4.1
    @backstage/types: 1.1.0
    @backstage/version-bridge: 1.0.4

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Turns out there was quite a bit of usage of unsupported TypeScript features in our configuration schemas. I founds some ways to make the config loader more strict and reject invalid schemas, rather than having them silently be ignored. This fixes #18812 and likely a couple of other related bugs. In general this is work towards #18176 too.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by sennyeya — APPROVED** [review]
Nice!

File: packages/config-loader/src/schema/collect.ts [hunk 1]
```diff
@@ -183,17 +184,40 @@ async function compileTsSchemas(paths: string[]) {
...
+        );
+      }
+
+      // This makes sure that no unsupported types are used in the schema, for example `Record<,>`.
+      // The generator will extract these as a schema reference, which will in turn be broken for our usage.
+      const reffedDefs = Object.keys(generator?.ReffedDefinitions ?? {});
+      if (reffedDefs.length !== 0) {
+        const lines = reffedDefs.join(`${EOL}  `);
+        throw new Error(
+          `Invalid configuration schema in ${path}, the following definitions are not supported:${EOL}${EOL}  ${lines}`,
```
Comment (adamdmharvey) [inline packages/config-loader/src/schema/collect.ts:218]: noted double EOL var; was that intentional?

**Review by freben — APPROVED** [review]
Amaze, cool that you found this

**Review by adamdmharvey — COMMENTED** [review]
<3

File: packages/config-loader/src/schema/collect.ts (hunk 1 above)
Comment (Rugvip) [reply packages/config-loader/src/schema/collect.ts:218]: Yep! To give a bit more spacing between the message itself and the horrible blobs of JSON :p

**Pull request merged**